pip_audit/_dependency_source/resolvelib/pypi_provider.py

```diff
@@ -418,35 +425,43 @@ def find_matches(
-                for candidate in all_candidates
-                if candidate.version not in bad_versions
-                # NOTE(ww): We use `filter(...)` instead of checking
-                # `candidate.version in r.specifier` because the former has subtle (and PEP 440
-                # mandated) behavior around prereleases. Specifically, `filter(...)`
-                # returns prereleases even if not explicitly configured, but only if
-                # there are no non-prereleases.
-                # See: https://github.com/pypa/pip-audit/issues/472
-                and all([any(r.specifier.filter((candidate.version,))) for r in requirements])
-                # HACK(ww): Additionally check that each candidate's name matches the
-                # expected project name (identifier).
-                # This technically shouldn't be required, but parsing distribution names
-                # from package indices is imprecise/unreliable when distribution filenames
-                # are PEP 440 compliant but not normalized.
-                # See: https://github.com/pypa/packaging/issues/527
-                and candidate.name == identifier
-            ],
-            key=attrgetter("version", "is_wheel"),
-            reverse=True,
-        )
+        candidates: list[Candidate]
+        try:
+            candidates = sorted(
+                [
+                    candidate
+                    for candidate in all_candidates
+                    if candidate.version not in bad_versions
+                    # NOTE(ww): We use `filter(...)` instead of checking
+                    # `candidate.version in r.specifier` because the former has subtle (and PEP 440
+                    # mandated) behavior around prereleases. Specifically, `filter(...)`
+                    # returns prereleases even if not explicitly configured, but only if
+                    # there are no non-prereleases.
+                    # See: https://github.com/pypa/pip-audit/issues/472
+                    and all([any(r.specifier.filter((candidate.version,))) for r in requirements])
+                    # HACK(ww): Additionally check that each candidate's name matches the
+                    # expected project name (identifier).
+                    # This technically shouldn't be required, but parsing distribution names
+                    # from package indices is imprecise/unreliable when distribution filenames
+                    # are PEP 440 compliant but not normalized.
+                    # See: https://github.com/pypa/packaging/issues/527
+                    and candidate.name == identifier
+                ],
+                key=attrgetter("version", "is_wheel"),
+                reverse=True,
+            )
+        except PyPINotFoundError as e:
+            skip_reason = str(e)
+            logger.debug(skip_reason)
+            self.skip_deps.append(SkippedDependency(name=identifier, skip_reason=skip_reason))
+            return
```

Comment:
@woodruffw I just found a serious problem here as I was fixing unit tests. The fact that we record the skipped candidate is fine, but then we proceed to return 0 candidates for a given requirement. This causes the resolver to raise a `ResolutionImpossible`.

In the past, we just skipped the top-level requirement and then try resolving the next one (since it was doing one at a time). But the resolver is "all or nothing" so we can't really get partial results from it. And once we get into `find_matches`, I don't think there's a way to backtrack and say "nevermind, just forget about that requirement I said I needed earlier". I'm at a bit of a loss figuring out how to fix this.

Reply:
Yeah, this is reminding me why I thought "one at a time" made sense despite not accurately reflecting the actual dependency resolution process 😅 

IMO a dummy candidate here is fine, but we should probably tread carefully -- it'd be good to have it be a separate type entirely similarly to how we handle `SkippedDependency` (`SkippedCandidate` maybe?).